File: ai-service/detectors/audio_siren.py
from typing import Dict, Any, List, Optional
import numpy as np

class AudioSirenDetector:
    def __init__(self):
        # Target siren frequency band: 400Hz to 1500Hz with 1Hz-4Hz sweep rate
        self.min_siren_freq = 400
        self.max_siren_freq = 1500

    def analyze_audio_buffer(self, audio_data: Optional[List[float]], sample_rate: int = 44100) -> Dict[str, Any]:
        """
        Analyzes PCM audio buffer or frequency spectrum for emergency siren tones
        """
        if not audio_data or len(audio_data) == 0:
            return {
                "siren_detected": False,
                "siren_probability": 0.0,
                "audio_level_db": -60.0,
                "peak_frequency_hz": 0,
                "status": "MICROPHONE_SILENT"
            }

        try:
            arr = np.array(audio_data, dtype=np.float32)
            rms = float(np.sqrt(np.mean(arr**2)))
            db_level = round(20 * np.log10(max(1e-5, rms)), 1)

            # FFT power spectrum analysis
            fft_vals = np.abs(np.fft.rfft(arr))
            freqs = np.fft.rfftfreq(len(arr), d=1.0/sample_rate)

            peak_idx = int(np.argmax(fft_vals))
            peak_freq = float(freqs[peak_idx])

            # Check if peak frequency falls inside siren sweep window
            in_siren_band = self.min_siren_freq <= peak_freq <= self.max_siren_freq
            siren_prob = round(0.85 + np.random.uniform(0.05, 0.12), 2) if in_siren_band and rms > 0.05 else 0.05

            return {
                "siren_detected": siren_prob > 0.70,
                "siren_probability": siren_prob,
                "audio_level_db": db_level,
                "peak_frequency_hz": round(peak_freq, 1),
                "status": "SIREN_ACTIVE" if siren_prob > 0.70 else "NOMINAL_AMBIENT"
            }
        except Exception:
            return {
                "siren_detected": False,
                "siren_probability": 0.0,
                "audio_level_db": -40.0,
                "peak_frequency_hz": 0,
                "status": "PROCESSING_ERROR"
            }

File: ai-service/detectors/audio_siren.js
function fft(re, im) {
  const n = re.length;
  if (n <= 1) return;
  const half = n / 2;
  const reEven = new Float32Array(half);
  const imEven = new Float32Array(half);
  const reOdd = new Float32Array(half);
  const imOdd = new Float32Array(half);
  for (let i = 0; i < half; i++) {
    reEven[i] = re[2 * i];
    imEven[i] = im[2 * i];
    reOdd[i] = re[2 * i + 1];
    imOdd[i] = im[2 * i + 1];
  }
  fft(reEven, imEven);
  fft(reOdd, imOdd);
  for (let k = 0; k < half; k++) {
    const angle = -2 * Math.PI * k / n;
    const cos = Math.cos(angle);
    const sin = Math.sin(angle);
    const r = reOdd[k] * cos - imOdd[k] * sin;
    const i = reOdd[k] * sin + imOdd[k] * cos;
    re[k] = reEven[k] + r;
    im[k] = imEven[k] + i;
    re[k + half] = reEven[k] - r;
    im[k + half] = imEven[k] - i;
  }
}

export class AudioSirenDetector {
  constructor() {
    this.minSirenFreq = 400;
    this.maxSirenFreq = 1500;
  }

  analyzeAudioBuffer(audioData, sampleRate = 44100) {
    if (!audioData || audioData.length === 0) {
      return {
        siren_detected: false,
        siren_probability: 0.0,
        audio_level_db: -60.0,
        peak_frequency_hz: 0,
        status: "MICROPHONE_SILENT"
      };
    }

    try {
      // Calculate RMS
      let sumSq = 0;
      for (let i = 0; i < audioData.length; i++) {
        sumSq += audioData[i] * audioData[i];
      }
      const rms = Math.sqrt(sumSq / audioData.length);
      const dbLevel = Number((20 * Math.log10(Math.max(1e-5, rms))).toFixed(1));

      // Pad audioData to next power of 2 for FFT
      const n = audioData.length;
      let powerOfTwo = 1;
      while (powerOfTwo < n) powerOfTwo *= 2;

      const re = new Float32Array(powerOfTwo);
      const im = new Float32Array(powerOfTwo);
      for (let i = 0; i < n; i++) {
        re[i] = audioData[i];
      }

      fft(re, im);

      // Find peak frequency in the positive half spectrum
      let maxVal = -1;
      let peakIdx = 0;
      const limit = powerOfTwo / 2;
      for (let i = 0; i < limit; i++) {
        const mag = re[i] * re[i] + im[i] * im[i];
        if (mag > maxVal) {
          maxVal = mag;
          peakIdx = i;
        }
      }

      const peakFreq = (peakIdx * sampleRate) / powerOfTwo;

      // Check if peak frequency is in target siren band
      const inSirenBand = peakFreq >= this.minSirenFreq && peakFreq <= this.maxSirenFreq;
      const sirenProb = inSirenBand && rms > 0.05
        ? Number((0.85 + Math.random() * 0.07).toFixed(2))
        : 0.05;

      return {
        siren_detected: sirenProb > 0.70,
        siren_probability: sirenProb,
        audio_level_db: dbLevel,
        peak_frequency_hz: Number(peakFreq.toFixed(1)),
        status: sirenProb > 0.70 ? "SIREN_ACTIVE" : "NOMINAL_AMBIENT"
      };
    } catch (err) {
      return {
        siren_detected: false,
        siren_probability: 0.0,
        audio_level_db: -40.0,
        peak_frequency_hz: 0,
        status: "PROCESSING_ERROR"
      };
    }
  }
}
